Entity company

chu4 opened the pull request:
Company model with controller and assosiations to user
+basic views (using simple_form for new and edit views)
+tests for the model and controller respectively(with factories)


![companies_edit](https://user-images.githubusercontent.com/23340152/34176293-27df9c30-e508-11e7-8967-cf4390ad87f9.PNG)
![companies_index](https://user-images.githubusercontent.com/23340152/34176294-2823392c-e508-11e7-9ecb-450a42286ba8.PNG)
![companies_new](https://user-images.githubusercontent.com/23340152/34176295-2855458e-e508-11e7-82c9-4fd08916db9c.PNG)
![companies_show](https://user-images.githubusercontent.com/23340152/34176296-28858384-e508-11e7-9dc3-ff40b467b8bf.PNG)

igorkasyanchuk commented:
move it all to "/account/" namespace as we discussed.

igorkasyanchuk commented on app/assets/javascripts/companies.js:
```diff
@@ -0,0 +1,2 @@
+// Place all the behaviors and hooks related to the matching controller here.
```
remove this file

chu4 replied:
removed as well as companies.scss

igorkasyanchuk commented on app/controllers/companies_controller.rb:
```diff
@@ -0,0 +1,53 @@
+  before_action :authenticate_user!
+
+  def index
+    @companies = Company.all
```
so as any user I can see all companies?

igorkasyanchuk commented on app/controllers/companies_controller.rb:
```diff
@@ -0,0 +1,53 @@
+  end
+
+  def current_company
+    Company.find(params[:id])
```
as any user, I can open any company? Just put a different ID in URL

PS: use current_user.companies....

igorkasyanchuk commented on config/routes.rb:
```diff
@@ -1,4 +1,6 @@
   devise_for :users
   # For details on the DSL available within this file, see http://guides.rubyonrails.org/routing.html
+
+  resources :companies
```
move to account namespace

chu4 replied:
Added account namespace and refactored routes, controller, views and tests with new path helpers

igorkasyanchuk commented on spec/controllers/companies_controller_spec.rb:
```diff
@@ -0,0 +1,151 @@
+require 'rails_helper'
+
+RSpec.describe CompaniesController, type: :controller do
+  let(:user) { create :user }
```
always use "let!" instead of "let"

chu4 replied:
changed it, but in this case it was 'let' cause user was only used in 5th line for signing_in before each test

igorkasyanchuk commented on spec/controllers/companies_controller_spec.rb:
```diff
@@ -0,0 +1,151 @@
+  end
+
+  describe "GET #show" do
+    let!(:company) { create :company }
```
create :company change to create(:company)
and in other places too

chu4 commented:
Improved code, respectively to the suggested changes

igorkasyanchuk commented on app/controllers/account/companies_controller.rb:
```diff
@@ -0,0 +1,55 @@
+module Account
+  class CompaniesController < ApplicationController
```
change to Account::CompaniesController

igorkasyanchuk replied:
and remove module Account

igorkasyanchuk commented on app/controllers/account/companies_controller.rb:
```diff
@@ -0,0 +1,55 @@
+    before_action :authenticate_user!
+
+    def index
+      @companies = Company.all
```
current_user.companies.all

igorkasyanchuk commented on app/controllers/account/companies_controller.rb:
```diff
@@ -0,0 +1,55 @@
+    end
+
+    def current_company
+      Company.find(params[:id])
```
current_user.companies.find

romaslmd commented on app/controllers/account/companies_controller.rb:
```diff
@@ -0,0 +1,55 @@
+    def update
+      @company = current_company
+      if @company.update_attributes(company_params)
+        redirect_to account_company_path
```
is this correct path? i believe for resources it should be either `account_companies_path` or `account_company_path(@company)`

romaslmd replied:
some flash message?

chu4 replied:
'account_company_path' works for updating but doesn't for creating, so there i'll use 'account_company_path(@company)' as suggested.
will add flashes, thanks

romaslmd commented on app/helpers/companies_helper.rb:
```diff
@@ -0,0 +1,2 @@
+module CompaniesHelper
```
incorrect module location
is it really needed?

chu4 replied:
true, as of the moment there is no need in that helper, so i removed it

romaslmd commented on spec/controllers/account/companies_controller_spec.rb:
```diff
@@ -0,0 +1,151 @@
+
+    it "populates an array of companies" do
+      get :index
+      expect(assigns(:companies)).to eq([company])
```
never do like this
`ActiveRecord:Association_Collection` is not an array
prefer to use `RSpec::Matchers#contain_exactly`

chu4 commented:
Refactored with suggested changes. Added Flash messages. Updated with current master branch.

flash message example
![image](https://user-images.githubusercontent.com/23340152/34222032-cf16b550-e5c2-11e7-9f31-9813fc7ccbd6.png)

igorkasyanchuk commented on app/views/layouts/application.html.slim:
```diff
@@ -11,4 +11,6 @@ html
       = notice
     p.alert
       = alert
```
what is "=alert"? above. And now you have added flash.each? remove line #13 probably

chu4 replied:
that line was generated by default, removed

igorkasyanchuk commented on spec/controllers/account/companies_controller_spec.rb:
```diff
@@ -0,0 +1,143 @@
+
+RSpec.describe Account::CompaniesController, type: :controller do
+  let!(:company) { create(:company) }
+  before(:each) { sign_in User.last }
```
why User.last? This is wrong. Use owner of a company.

igorkasyanchuk replied:
and remember ... in many cases when you have code like ".first", ".last" there is a bug

chu4 replied:
My bad, company.user should be a better choice

igorkasyanchuk commented on spec/models/company_spec.rb:
```diff
@@ -0,0 +1,16 @@
+
+RSpec.describe Company, type: :model do
+  describe 'validations with factories' do
+    let(:val_company) { build :company }
```
use "let!" always
and "build(:company)" with brackets
